### PR TITLE
MosaicBERT Migration

### PR DESCRIPTION
**Pull request by cojennin**
This PR migrates MosaicBERT ([blogpost](https://www.mosaicml.com/blog/mosaicbert), [workshop paper](https://openreview.net/pdf?id=WH1S0gonzR)) from `mosaicml/examples` --> `mosaicml/llmfoundry`. The goal of the PR is to replicate the functionality of MosaicBERT pretraining and finetuning with minimal changes to the MosaicBERT code.

This addition is significant, as it expands the ambit of the llmfoundry repo beyond autoregressive models and the MPT architecture. This also lays the groundwork for BERT-like embedding models.

In the original [`mosaicml/examples`](https://github.com/mosaicml/examples/tree/main/examples/benchmarks/bert) repo, you can pretrain both BERT and MosaicBERT. You can also finetune composer checkpoints of BERT and MosaicBERT on the GLUE benchmark. This port maintains these capabilities.

## File Structure

_Here's how the files have roughly been copied over:_

| mosaicml/examples/benchmarks/ | mosaicml/llm-foundry|
|-------|-------|
| bert/src/bert_layers.py | llmfoundry/layers/mosaicbert_layers.py |
| bert/src/bert_padding.py | llmfoundry/models/utils/bert_padding.py |
| bert/src/hf_bert.py | llmfoundry/models/hf/hf_bert.py |
| bert/src/mosaic_bert.py | llmfoundry/layers/mosaicbert/modeling_mosaicbert.py |
|bert/src/configuation_bert.py | llmfoundry/layers/mosaicbert/config_mosaicbert.py |
| bert/tests/... (6 files) | llm-foundry/tests/ ... (6 files) |

### Pretraining

Both HF BERT and MosaicBERT pretraining can be run with 
```
cd llm-foundry/scripts
composer train/train.py /mnt/config/parameters.yaml
```

* Within the `llmfoundry/models` folder, we added `hf_bert.py` to the `hf` subfolder and additionally create a new folder `mosaicbert` with the files `configuration_mosaicbert.py` and `modeling_mosaicbert.py`.

* We also added `ComposerMosaicBertForMaskedLM`, `ComposerMosaicBertForSequenceClassification`, `ComposerHFBertForMaskedLM` and `ComposerHFBertForSequenceClassification` to the `model_registry.py`

* Note that we also changed `ComposerBertForMaskedLM` to `ComposerMosaicBertForMaskedLM` to make it clear exactly when we use MosaicBERT versus the classic BERT.
* We changed `model_cofig` to `resolved_om_model_config`
* `LanguageCrossEntropy(ignore_index=-100, vocab_size=model.config.vocab_size)` was changed to `LanguageCrossEntropy(ignore_index=-100)` for updated Composer compatibility
* `mosaicbert_layers.py` is kept in the layers folder instead of the `models/mosaicbert` folder. The mosaicbert_layers are modified versions of the original transformers BERT layers, and this structure is slightly different from how we have designed the layers for llm-foundry. In order to change as few files as possible, we have kept the `mosaicbert_layers.py` the same.

### Finetuning

In the `scripts` folder, we specify the files necessary for running GLUE finetuning in the [`train/finetune_bert_glue/`](https://github.com/cojennin/llm-foundry/tree/bert-v0/scripts/train/finetune_bert_glue).

Instead of creating the function `create_hf_bert_classification()`, as in the mosaicml/examples repo, we simply define `build_composer_model(model_cfg, tokenizer)` to match the model name (e.g. `mosaicbert_masked_lm`, `hf_bert_masked_lm` etc.) to the model registry.

A few other small changes were made to `glue.py`.

## Experiments

- [x] HuggingFace BERT pretraining ([wandb](https://wandb.ai/mosaic-ml/mbed-test) and [wandb](https://wandb.ai/mosaic-ml/bert-llmfoundry-port))

- [x] MosaicBERT pretraining ([wandb](https://wandb.ai/mosaic-ml/bert-llmfoundry-port))
- [x] HuggingFace BERT finetuning on 1 GPU ([wandb](https://wandb.ai/mosaic-ml/bert-glue-test?workspace=user-jportes))
- [ ] HuggingFace BERT finetuning on 8 GPUs (wandb)
- [ ] MosaicBERT finetuning on 1 GPU (wandb)
- [ ] MosaicBERT finetuning on 8 GPUs (wandb)

## TO DO
- [ ] port over `test_glue.py`, `test_classification.py`, `test_main.py`

## Nice to Have
- [ ] README for pretraining
- [ ] README for finetuning

### Potential To Dos once this PR is approved:
* The logic in class `ComposerMosaicBertForMaskedLM(HuggingFaceModel)` could be consolidated, because now it inherits from `HuggingFaceModel`

### Other changes:

* `pretrained_checkpoint = resolved_om_model_config.get('pretrained_checkpoint')` was removed from `hf_bert.py`

### CODE REVIEW

**Comment by jacobfulano** [comment]
@dakinggg @alextrott16 this PR is ready for review

**Review by alextrott16 — CHANGES_REQUESTED** [review]
Overall, this looks very solid. I do think some YAMLs got overlooked and need to be updated to reflect how everything is landing here in `llm-foundry`. I think that needs to be addressed before stamping Approve.

Also, please weigh in on the "signature" of the model constructors, ie the config structure that the new BERT model constructors expect. I'm thinking they should be as unified as possible with the existing constructors. If unifying seems dumb, please say so. Or, if it's just out of scope for now, then I'm OK with scoping a refactor into a future PR.

Finally, let's get all the tests/sanity checks cleared before we merge. For the record, the PR summary (which includes that checklist) is absolutely fantastic and is so appreciated by this humble reviewer!

File: scripts/train/yamls/pretrain/mosaicbert-base-uncased.yaml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+# Model
+model:
+  name: mosaicbert_masked_lm
+  pretrained_model_name: ${tokenizer_name}
+  tokenizer_name: ${tokenizer_name}
+  # Mosaic BERT 'base' generally uses the default architecture values for from the Hugging Face BertConfig object
+  # Note: if using the pretrained_checkpoint argument to create a model from an existing checkpoint, make sure
+  # the model_config settings match the architecture of the existing model
+  model_config:
```
Comment (dakinggg) [inline scripts/train/yamls/pretrain/mosaicbert-base-uncased.yaml:24]: same

**Review by dakinggg — COMMENTED** [review]
A couple high level comments. We need tests with this PR and we need documentation with this PR. I suggest a small section of train readme about bert, and a readme in `finetune_bert_glue`

**Comment by dakinggg** [comment]
@cojennin should we close this?

**Comment by dakinggg** [comment]
Closing as this is not being actively worked on.

**Pull request closed without merging**